### PR TITLE
KIN-338 functionality to create certificates with files

### PR DESCRIPTION
**Pull request by FernandoJoachin**
[KIN-338](https://loopcrack.atlassian.net/browse/KIN-338) Small [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=loopcrack-org&utm_campaign=badge_info)

### Problem

The logic of creating certificate was integrated without the file functionality.

### Solution

Integrate files functionality for the certificates files

### Requirements checklist

- [x] It must validate that on the certificate file input it is only uploaded pdf files
- [x] It must validate that on the certificate preview input it is only uploaded png/jpg files
- [x] If it passes all the validations, the files must be stored on the server and the file routes as must be stored in the database

### Testing

- [x] Go to /admin/certificados/crear and create a certificate
- [x] In the Preview Image field, upload a file that is not an image or that exceeds the specified weight or size
- [x] In the Certificate File field, upload a file that is not a pdf or that exceeds the specified weight

[KIN-338]: https://loopcrack.atlassian.net/browse/KIN-338?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by CheloChidoone — CHANGES_REQUESTED** [review]
Good job @FernandoJoachin it works well but I found something that we should change that affects this PR, when we use an image that exceeds the dimensions you specified we get this message: 
image

I think this message makes it seem like the image should be exactly of those dimensions when in reality those are the max dimensions it could have so the message should say something like `no debe exceder estas dimensiones`.

@Cristianpan @Mauriciocr207 I don't know if these change should be made in this PR but it affects it and any other place where the image dimensions will be checked, please let us know your toughts about these change.

**Comment by Mauriciocr207** [comment]
> [Dec 14, 2023](#event-11247227267)
> [CheloChidoone](/CheloChidoone)
> ** [CheloChidoone](/CheloChidoone) ** requested changes [ Dec 14, 2023 ](#pullrequestreview-1782324773)
> [ View reviewed changes ](/loopcrack-org/Kinub/pull/151/files/7fd47978e6c1d7a09fc2c5a94b2a38b811a31918)

About this, I think that could be changed here due that is a little change

File: app/Controllers/CtrlCertificate.php [hunk 1]
```diff
@@ -2,13 +2,27 @@
...
 
-class CtrlCertificate extends BaseController
+class CtrlCertificate extends CtrlApiFiles
 {
+    protected FileValidationConfig $fileConfig;
+
+    public function __construct()
+    {
+        $fileConfigBuilder = new FileValidationConfigBuilder('/admin/certificados');
+        $fileConfigBuilder->builder('certificatePreviewId')->minFiles(1)->maxFiles(1)->maxSize(2, 'MB')->isImage()->maxDims(1000, 1500)->build();
```
Comment (LUGAMAFE) [inline app/Controllers/CtrlCertificate.php:20]: Why `certificatePreviewId` should be only `certificatePreview`?
And i think the max dims are to much, look a real example
```suggestion
        $fileConfigBuilder->builder('certificatePreview')->minFiles(1)->maxFiles(1)->maxSize(2, 'MB')->isImage()->maxDims(700, 700)->build();
```

image

File: app/Controllers/CtrlCertificate.php [hunk 2]
```diff
@@ -2,13 +2,27 @@
...
+        $fileConfigBuilder->builder('certificatefileId')->minFiles(1)->maxFiles(1)->maxSize(30, 'MB')->isPDF()->build();
```
Comment (LUGAMAFE) [inline app/Controllers/CtrlCertificate.php:21]: Same why `certificatefileId` and not  just `certificatefile`

File: app/Views/admin/certificates/CertificateCreate.php [hunk 3]
```diff
@@ -23,13 +23,17 @@
                         <i class="ri-arrow-left-fill label-icon align-middle rounded-pill fs-16 me-2"></i>Volver
                     </a>
                 </div>
-                <form id="createproduct-form" autocomplete="off" class="needs-validation" method="POST">
+                <form id="form" autocomplete="off" class="needs-validation" novalidate method="POST">
```
Comment (LUGAMAFE) [inline app/Views/admin/certificates/CertificateCreate.php:26]: Why you added novalidate?
```suggestion
                <form id="form" autocomplete="off" class="needs-validation" method="POST">
```

File: app/Views/admin/certificates/certificatesFormTemplate.php [hunk 4]
```diff
@@ -31,7 +31,7 @@
             </div><!-- end card header -->
 
             <div class="card-body">
-                <input type="file" class="filepond" name="certificatePreviewId">
+                <?= view('admin/components/inputFilePond', ['config' => $filepondConfig['certificatePreviewId']]) ?>
```
Comment (LUGAMAFE) [inline app/Views/admin/certificates/certificatesFormTemplate.php:34]:
```suggestion
                <?= view('admin/components/inputFilePond', ['config' => $filepondConfig['certificatePreview']]) ?>
```

File: app/Views/admin/certificates/certificatesFormTemplate.php [hunk 5]
```diff
@@ -46,7 +46,7 @@
             </div><!-- end card header -->
 
             <div class="card-body">
-                <input type="file" class="filepond" name="certificatefileId">
+                 <?= view('admin/components/inputFilePond', ['config' => $filepondConfig['certificatefileId']]) ?>
```
Comment (LUGAMAFE) [inline app/Views/admin/certificates/certificatesFormTemplate.php:49]:
```suggestion
                 <?= view('admin/components/inputFilePond', ['config' => $filepondConfig['certificatefile']]) ?>
```

**Review by LUGAMAFE — CHANGES_REQUESTED** [review]
Nits @FernandoJoachin

**Review by CarlosMay7 — COMMENTED** [review]
I think it looks good but I have a little problem, I don't know if it's just me but the first time I try to create a certificate without a name it redirects me to a 404 page

[Demo](https://drive.google.com/file/d/1e49iYOOfLBROKowiak6cnv4rDU9CvCGx/view?usp=sharing)

**Comment by FernandoJoachin** [comment]
> I think it looks good but I have a little problem, I don't know if it's just me but the first time I try to create a certificate without a name it redirects me to a 404 page
> 
> [Demo](https://drive.google.com/file/d/1e49iYOOfLBROKowiak6cnv4rDU9CvCGx/view?usp=sharing)

Has this happened to anyone else? Because I've checked and it works for me.

**Review by Cristianpan — APPROVED** [review]
Please considere that comment that I've left you and thanks to make the other changes.

File: app/Views/admin/certificates/certificatesFormTemplate.php [hunk 6]
```diff
@@ -9,7 +9,7 @@
             </div>
             <div class="card-body">
                 <div>
-                    <input type="text"  class="form-control <?= isset($errors['certificatefileName']) ? 'is-invalid' : '' ?>" name="certificatefileName" id="certificatefileName" value="<?= old('certificatefileName') ?? $certificate['certificatefileName'] ?? ''; ?>" placeholder="Ingrese el nombre" required>
+                    <input type="text"  class="form-control <?= isset($errors['certificatefileName']) ? 'is-invalid' : '' ?>" name="certificatefileName" id="certificatefileName" value="<?= old('certificatefileName') ?? $certificate['certificatefileName'] ?? ''; ?>" placeholder="Ingrese el nombre">
```
Comment (Cristianpan) [inline app/Views/admin/certificates/certificatesFormTemplate.php:12]: Please add required attribute in this input

**Comment by Mauriciocr207** [comment]
> I think it looks good but I have a little problem, I don't know if it's just me but the first time I try to create a certificate without a name it redirects me to a 404 page
> 
> [Demo](https://drive.google.com/file/d/1e49iYOOfLBROKowiak6cnv4rDU9CvCGx/view?usp=sharing)

I don´t know yet why this happened you, but it seems like the method patch with we load files fails. This still happen to you?

**Review by Mauriciocr207 — APPROVED** [review]
Good job, @FernandoJoachin!. Same comments as @Cristianpan . And if you made the previous corrections, please request changes again.

**Comment by LUGAMAFE** [comment]
> I think it looks good but I have a little problem, I don't know if it's just me but the first time I try to create a certificate without a name it redirects me to a 404 page
> 
> [Demo](https://drive.google.com/file/d/1e49iYOOfLBROKowiak6cnv4rDU9CvCGx/view?usp=sharing)

Hey @CarlosMay7 can you check that your CI_ENVIRONMENT is in production mode.

image

**Review by LUGAMAFE — CHANGES_REQUESTED** [review]
ignore this

**Review by LUGAMAFE — APPROVED** [review]
Nice @FernandoJoachin. Same comments as @Cristianpan please before merge

**Review by CheloChidoone — APPROVED** [review]

**Review by CarlosMay7 — APPROVED** [review]

**Pull request merged**